blog/_src/posts/2017-06-05-syntactic-parametricity-strikes-again.md

```diff
@@ -0,0 +1,399 @@
+only present the reasoning informally, but it can easily be made
+formally precise.
+
+We start by looking at the shape of the search space for
+
+\\[
+    \judge{}{∀α. \List α → \List α}
+\\]
+or, said otherwise, of the judgment
+\\[
+    \judge{}{\List α → \List α}
+\\]
+
+with a fresh/abstract type variable \\(α\\). (I will not be keeping
+opened type variables in context to avoid confusing them with hypotheses.)
+
+Any derivation of a function type, without loss of generality
+(w.l.o.g), is equivalent to a derivation starting with a function
+introduction. This is the η-expansion rule for functions: any proof
+term \\(e\\) is equivalent to \\(λx.~(e~x)\\), a proof that starts
+with a \\(λ\\). So any proof can be taken to start as follows:
+\\[
+\rule{
+\judge{\List \alpha}{\List \alpha}
+}{
+\judge{}{\List \alpha \to \List \alpha}
+}
+\\]
+we can, w.l.o.g, unfold the recursive type in the context
+(\\(\List α = 1 + (α × \List α)\\)):
+\\[
+\rule{
+\judge{1 + (α × \List α)}{\List α}
+}{
+\rule{
+\judge{\List α}{\List α}
+}{
+\judge{}{\List α → \List α}
+}}
+\\]
+
+A derivation with a sum type as hypothesis can, w.l.o.g, be assumed to
+start by splitting on this pair (this is the η-expansion rule
+for sums):
+\\[
+\rule{
+\judge{1}{\List α}
+\quad
+\judge{α × \List α}{\List α}
+}{
+\rule{
+\judge{1 + (α × \List α)}{\List α}
+}{
+\rule{
+\judge{\List α}{\List α}
+}{
+\judge{}{\List α → \List α}
+}}}
+\\]
+
+In the right subgoal, we can always, w.l.o.g, split a hypothesis of
+product type:
+\\[
+\rule{
+\emptyrule{\judge{1}{\List α}}
+\quad
+\rule{
+\judge{α, \List α}{\List α}
+}{
+\judge{α × \List α}{\List α}
+}}{
+\rule{
+\judge{1 + (α × \List α)}{\List α}
+}{
+\rule{
+\judge{\List α}{\List α}
+}{
+\judge{}{\List α → \List α}
+}}}
+\\]
+
+Now, an interesting pattern emerges. In the process of trying to prove
+\\(\judge{\List α}{\List α}\\), we have to prove the (right) subgoal
+\\(\judge{α,\List α}{α}\\). We can generalize this derivation by assuming that we
+start with some number \\(n\\) of variables of type \\(α\\) in the context
+(we write \\(α^n\\) for this):
+\\[
+\rule{
+\rule{
+\judge{\alpha^n}{\List \alpha}
+}{
+\judge{\alpha^n, 1}{\List \alpha}
+}
+\quad
+\rule{
+\judge{\alpha^{n+1}, \List \alpha}{\List \alpha}
+}{
+\judge{\alpha^n, \alpha \times \List \alpha}{\List \alpha}
+}}{
+\rule{
+\judge{\alpha^n, 1 + (\alpha \times \List \alpha)}{\List \alpha}
+}{
+\judge{\alpha^n, \List \alpha}{\List \alpha}
+}}
+\\]
+
+\\[
+\newcommand{\llbracket}{〚}
+\newcommand{\rrbracket}{〛}
+\newcommand{\sem}[1]{\llbracket{} #1 \rrbracket{}}
+\\]
+
+Let us write \\(\sem{\judge{\alpha^n, \List \alpha}{\List \alpha}}\\)
+for the search space corresponding to all possible derivations of the
+judgment \\(\judge{\alpha^n, \List \alpha}{\List \alpha}\\).  All the
+proof steps above have been done "without loss of generality"
+(in terms of focusing, we only used invertible rules), so they appear
```

Comment:
if you do remove the `w.l.o.g.`, then here's one suggestion for rewording this:

> All the proof steps above appear in any such derivation (in terms of focusing, this is because the steps used only invertible rules)